### PR TITLE
Add a source property to Warnings

### LINKED ISSUE #159: Add a source property to Warnings
I would like to a new `source` property to warnings to be able to identity who created the warning. Nowadays warnings can be generated from JS-X-Ray or Scanner (take [invalid-semver warning](https://github.com/NodeSecure/scanner/blob/bf36fb0c44df63b00dc2662afe79aefaa0fd3689/src/utils/semver.js#L51)).

The task consist of updating the TS interface and add the source in the [generateWarning](https://github.com/NodeSecure/js-x-ray/blob/master/src/warnings.js#L56) function.

https://github.com/NodeSecure/js-x-ray/blob/c91a37e8e530b72320c497c2beae475b3d00415d/types/warnings.d.ts#L24-L32

### PR DESCRIPTION
**Pull request by fabnguess**
The purpose of this PR is to resolve #159 

close #159

### CODE REVIEW

File: src/warnings.js [hunk 1]
```diff
@@ -54,12 +54,12 @@ export const warnings = Object.freeze({
 });
 
 export function generateWarning(kind, options) {
-  const { location, file = null, value = null } = options;
+  const { location, file = null, value = null, source = "JS-X-Ray"} = options;
```
Comment (PierreDemailly) [inline src/warnings.js]: Should we validate source ?
```js
const kWarningsSource = new Set(["JS-X-Ray", "Scanner"]);
// [...]
if (!kWarningsSource.has(source) {
  throw new Error(`Source ${source} is not allowed`);
}
```

Comment (fabnguess) [reply src/warnings.js]: Sorry, I hadn't seen that before my last commit. I've finally removed the union type (should I leave it in place?).

Comment (PierreDemailly) [reply src/warnings.js]: I think it's ok to not check the source then 👍 
However, i would not set it as `null` by default but `JS-X-Ray` instead.

Comment (fabnguess) [reply src/warnings.js]: OK noted

File: types/warnings.d.ts [hunk 2]
```diff
@@ -25,6 +25,7 @@ interface WarningDefault {
   kind: WarningName;
   file?: string;
   value: string;
+  source?: string;
```
Comment (fraxken) [inline types/warnings.d.ts]: should be a mandatory property

File: src/warnings.js [hunk 3]
```diff
@@ -54,11 +54,11 @@ export const warnings = Object.freeze({
 });
 
 export function generateWarning(kind, options) {
-  const { location, file = null, value = null } = options;
+  const { location, file = null, value = null, source = "JS-X-Ray" } = options;
 
   if (kind === "encoded-literal") {
     return Object.assign(
-      { kind, value, location: [utils.toArrayLocation(location)] },
+      { kind, value, location: [utils.toArrayLocation(location)], source },
```
Comment (fraxken) [inline src/warnings.js:61]: You need to add source to the `result` Object too (few lines below).

**Review by fraxken — CHANGES_REQUESTED** [review]

File: src/warnings.js (hunk 1 above)
Comment (fraxken) [reply src/warnings.js]: I dont think we need to validate source (could limit customization of source in future).

File: src/warnings.js (hunk 3 above)
Comment (PierreDemailly) [reply src/warnings.js:61]: @fabnguess you missed the `result` const line 66

**Review by PierreDemailly — APPROVED** [review]

**Review by fraxken — APPROVED** [review]

**Comment by fraxken** [comment]
@allcontributors please add @fabnguess for code

**Pull request merged**